File: modules/carbontrack-pow/ios/CarbonTrackPowModule.swift
import ExpoModulesCore
import CryptoKit
import Foundation

public class CarbonTrackPowModule: Module {
  private let lock = NSLock()
  private var cancellations: [String: Bool] = [:]

  public func definition() -> ModuleDefinition {
    Name("CarbonTrackPow")

    AsyncFunction("solve") { (
      challenge: String,
      difficulty: Int,
      maxAttempts: Int,
      timeoutMs: Int,
      operationId: String
    ) -> [String: Any] in
      return try self.solve(
        challenge: challenge,
        difficulty: difficulty,
        maxAttempts: maxAttempts,
        timeoutMs: timeoutMs,
        operationId: operationId
      )
    }

    Function("cancel") { (operationId: String) in
      self.cancel(operationId: operationId)
    }
  }

  private func solve(
    challenge: String,
    difficulty: Int,
    maxAttempts: Int,
    timeoutMs: Int,
    operationId: String
  ) throws -> [String: Any] {
    if challenge.isEmpty || difficulty < 1 || difficulty > 256 || maxAttempts < 1 || timeoutMs < 1 || operationId.isEmpty {
      throw NSError(domain: "CarbonTrackPow", code: 1, userInfo: [
        NSLocalizedDescriptionKey: "Invalid proof-of-work challenge"
      ])
    }

    beginCancellation(operationId: operationId)
    let startedAt = DispatchTime.now().uptimeNanoseconds
    let timeoutNanos = UInt64(timeoutMs) * 1_000_000
    let prefix = Data("\(challenge):".utf8)
    var message = Data()
    message.reserveCapacity(prefix.count + String(maxAttempts).count)
    message.append(prefix)

    defer {
      removeCancellation(operationId: operationId)
    }

    for nonce in 0..<maxAttempts {
      if (nonce & 0x3ff) == 0 {
        if isCancelled(operationId: operationId) {
          throw NSError(domain: "CarbonTrackPow", code: 2, userInfo: [
            NSLocalizedDescriptionKey: "Proof-of-work calculation cancelled"
          ])
        }

        let elapsedNanos = DispatchTime.now().uptimeNanoseconds - startedAt
        if elapsedNanos > timeoutNanos {
          throw NSError(domain: "CarbonTrackPow", code: 3, userInfo: [
            NSLocalizedDescriptionKey: "Proof-of-work calculation timed out"
          ])
        }
      }

      if message.count > prefix.count {
        message.count = prefix.count
      }
      appendDecimalBytes(nonce, to: &message)
      let hash = SHA256.hash(data: message)
      if hasLeadingZeroBits(hash, difficulty: difficulty) {
        let elapsedMs = Int((DispatchTime.now().uptimeNanoseconds - startedAt) / 1_000_000)
        return [
          "nonce": String(nonce),
          "attempts": nonce + 1,
          "elapsedMs": elapsedMs
        ]
      }
    }

    throw NSError(domain: "CarbonTrackPow", code: 4, userInfo: [
      NSLocalizedDescriptionKey: "Proof-of-work attempt limit exceeded"
    ])
  }

  private func appendDecimalBytes(_ value: Int, to data: inout Data) {
    if value == 0 {
      data.append(48)
      return
    }

    var divisor = 1
    while value / divisor >= 10 {
      divisor *= 10
    }

    var remaining = value
    while divisor > 0 {
      let digit = remaining / divisor
      data.append(UInt8(48 + digit))
      remaining %= divisor
      divisor /= 10
    }
  }

  private func hasLeadingZeroBits<D: Sequence>(_ digest: D, difficulty: Int) -> Bool where D.Element == UInt8 {
    let bytes = Array(digest)
    if difficulty < 1 || difficulty > bytes.count * 8 {
      return false
    }

    let fullBytes = difficulty / 8
    for index in 0..<fullBytes {
      if bytes[index] != 0 {
        return false
      }
    }

    let remainingBits = difficulty % 8
    if remainingBits == 0 {
      return true
    }

    let mask = UInt8((0xff << (8 - remainingBits)) & 0xff)
    return (bytes[fullBytes] & mask) == 0
  }

  private func setCancelled(_ cancelled: Bool, operationId: String) {
    lock.lock()
    cancellations[operationId] = cancelled
    lock.unlock()
  }

  private func beginCancellation(operationId: String) {
    lock.lock()
    if cancellations[operationId] == nil {
      cancellations[operationId] = false
    }
    lock.unlock()
  }

  private func cancel(operationId: String) {
    setCancelled(true, operationId: operationId)
  }

  private func isCancelled(operationId: String) -> Bool {
    lock.lock()
    let cancelled = cancellations[operationId] ?? false
    lock.unlock()
    return cancelled
  }

  private func removeCancellation(operationId: String) {
    lock.lock()
    cancellations.removeValue(forKey: operationId)
    lock.unlock()
  }
}

File: modules/carbontrack-pow/android/src/main/java/expo/modules/carbontrackpow/CarbonTrackPowModule.kt
package expo.modules.carbontrackpow

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

class CarbonTrackPowModule : Module() {
  private data class CancellationState(
    val cancelled: AtomicBoolean,
    val createdAtNanos: Long
  )

  private val cancellations = ConcurrentHashMap<String, CancellationState>()
  private val cancellationTtlNanos = 5L * 60L * 1_000_000_000L

  override fun definition() = ModuleDefinition {
    Name("CarbonTrackPow")

    AsyncFunction("solve") {
      challenge: String,
      difficulty: Int,
      maxAttempts: Int,
      timeoutMs: Int,
      operationId: String ->
      solve(challenge, difficulty, maxAttempts, timeoutMs, operationId)
    }

    Function("cancel") { operationId: String ->
      if (operationId.isNotBlank()) {
        val now = System.nanoTime()
        pruneStaleCancellations(now)
        cancellations.compute(operationId) { _, existing ->
          existing?.also { it.cancelled.set(true) } ?: CancellationState(AtomicBoolean(true), now)
        }
      }
    }
  }

  private fun solve(
    challenge: String,
    difficulty: Int,
    maxAttempts: Int,
    timeoutMs: Int,
    operationId: String
  ): Map<String, Any> {
    if (challenge.isBlank() || difficulty < 1 || difficulty > 256 || maxAttempts < 1 || timeoutMs < 1 || operationId.isBlank()) {
      throw IllegalArgumentException("Invalid proof-of-work challenge")
    }

    pruneStaleCancellations()
    val cancellation = cancellations.computeIfAbsent(operationId) {
      CancellationState(AtomicBoolean(false), System.nanoTime())
    }
    val startedAt = System.nanoTime()
    val timeoutNanos = timeoutMs.toLong() * 1_000_000L
    val prefix = "$challenge:".toByteArray(StandardCharsets.UTF_8)
    val digest = MessageDigest.getInstance("SHA-256")
    val nonceBuffer = ByteArray(maxAttempts.toString().length)

    try {
      for (nonce in 0 until maxAttempts) {
        if ((nonce and 0x3ff) == 0) {
          if (cancellation.cancelled.get()) {
            throw IllegalStateException("Proof-of-work calculation cancelled")
          }
          if (System.nanoTime() - startedAt > timeoutNanos) {
            throw IllegalStateException("Proof-of-work calculation timed out")
          }
        }

        digest.reset()
        digest.update(prefix)
        val nonceLength = writeDecimalBytes(nonce, nonceBuffer)
        digest.update(nonceBuffer, 0, nonceLength)
        val hash = digest.digest()
        if (hasLeadingZeroBits(hash, difficulty)) {
          return mapOf(
            "nonce" to nonce.toString(),
            "attempts" to nonce + 1,
            "elapsedMs" to ((System.nanoTime() - startedAt) / 1_000_000L)
          )
        }
      }

      throw IllegalStateException("Proof-of-work attempt limit exceeded")
    } finally {
      cancellations.remove(operationId)
    }
  }

  private fun writeDecimalBytes(value: Int, buffer: ByteArray): Int {
    if (value == 0) {
      buffer[0] = '0'.code.toByte()
      return 1
    }

    var divisor = 1
    while (value / divisor >= 10) {
      divisor *= 10
    }

    var remaining = value
    var length = 0
    while (divisor > 0) {
      val digit = remaining / divisor
      buffer[length] = ('0'.code + digit).toByte()
      length += 1
      remaining %= divisor
      divisor /= 10
    }

    return length
  }

  private fun hasLeadingZeroBits(bytes: ByteArray, difficulty: Int): Boolean {
    if (difficulty < 1 || difficulty > bytes.size * 8) {
      return false
    }

    val fullBytes = difficulty / 8
    for (index in 0 until fullBytes) {
      if ((bytes[index].toInt() and 0xff) != 0) {
        return false
      }
    }

    val remainingBits = difficulty % 8
    if (remainingBits == 0) {
      return true
    }

    val mask = (0xff shl (8 - remainingBits)) and 0xff
    return ((bytes[fullBytes].toInt() and 0xff) and mask) == 0
  }

  private fun pruneStaleCancellations(nowNanos: Long = System.nanoTime()) {
    for ((operationId, state) in cancellations) {
      if (nowNanos - state.createdAtNanos > cancellationTtlNanos) {
        cancellations.remove(operationId, state)
      }
    }
  }
}
